### ISSUE TITLE
We need "sender" and "receiver" stats, not "track" stats

### ISSUE THREAD

**Issue by jan-ivar** [w3c/webrtc-stats#231]
[RTCMediaStreamTrackStats](https://w3c.github.io/webrtc-stats/#dom-rtcmediastreamtrackstats) sets up a nice symmetry between local fields:
```js
    unsigned long       framesCaptured;
    unsigned long       framesSent;
```
and remote fields:
```js
    unsigned long       framesReceived;
    unsigned long       framesDecoded;
    unsigned long       framesDropped;
    unsigned long       framesCorrupted;
    unsigned long       partialFramesLost;
    unsigned long       fullFramesLost;
```
Would you expect to be able to compare `framesSent` and `framesReceived`?

Unfortunately, this was designed before transceivers, senders and replaceTrack, so there's an impedance mismatch: In the new model, remote tracks are tied to *senders*, not local tracks.

E.g. take the ["Hold" example](http://w3c.github.io/webrtc-pc/#hold-functionality) in the spec:
 1. I put you on hold using `audio.sender.replaceTrack(null)`,
 2. and then I bring you back from hold with `audio.sender.replaceTrack(micTrack)`

Would you still expect to be able to compare `framesSent` and `framesReceived`?

We can't because the *local* stats have now been reset from the time of the second `replaceTrack`, whereas the remote stats are *never reset*. The counters are out of sync.

There's also stats bloat (on the local side only), in that you'll see one additional `detached` RTCMediaStreamTrackStats record for each time `replaceTrack(micTrack)` was called. In theory, the sum of all the `framesSent` with the same track `id` should be comparable to `framesReceived`, but that's quite a bit of work, asymmetrical and unintuitive.

Instead, we think people care more about transmission data in relation to the *sender*, not the track.

### Proposal

After splitting up the dictionary in https://github.com/w3c/webrtc-stats/issues/230, replace the `"track"` [RTCStatsType](https://w3c.github.io/webrtc-stats/#dom-rtcstatstype) with `"sender"` and `"receiver"` and have those refer to the local and remote half of this dictionary respectively. e.g.

s/LocalVideoStreamTrackStats/VideoSenderStats/
s/RemoteVideoStreamTrackStats/VideoReceiverStats/
s/LocalAudioStreamTrackStats/AudioSenderStats/
s/RemoteAudioStreamTrackStats/AudioReceiverStats/

**Comment by vr000m** [comment]
I think the local track stats make sense in this case to know if this are being sent or not, so maybe having both the track accessor and the sender accessor might make sense in this particular scenario. So sender and receiver object might make sense.

Although splitting the audio and video seems unnecessary (although visually appealing), because the sender and receiver would report only the appropriate stats.

**Comment by jan-ivar** [comment]
@vr000m Happy to hear it makes sense!

> Do the sender track stats in the case of hold keep increasing (across all implementations)?

Well, stats report on implementation performance, which is generally not spec'ed, and why we have stats. We see that for instance in [sending a disabled track](https://blog.mozilla.org/webrtc/fiddle-week-spec-compliant-getstats/), where some browsers send 1 frame per second for video, while others send 30.

That said, "hold" in the spec example renegotiates to "inactive" or "recvonly", so stats would reflect however browsers implement that. I believe they're barred from sending frames in that case, so you should see `framesSent` stop. But would `framesCaptured` stop? [It](https://w3c.github.io/webrtc-stats/#dom-rtcmediastreamtrackstats-framescaptured) doesn't even say if it stops for `!track.enabled`, so that seems out of scope / implementation dependent.

> I think the local track stats make sense in this case to know if this are being sent or not

I believe "sender" should tell you the same thing, assuming you know what you're sending.

**Comment by jan-ivar** [comment]
Though after `sender.replaceTrack(null)` it totally makes sense that `framesCaptured` should stop.

**Comment by jan-ivar** [comment]
> Although splitting the audio and video seems unnecessary (although visually appealing), because the sender and receiver would report only the appropriate stats.

if you're referring to https://github.com/w3c/webrtc-stats/issues/230 then that split is editorial only, not js-observable. But let's discuss it there.

**Comment by jan-ivar** [comment]
@alvestrand PTAL. With this in place I think the need for `detached` track stats goes away.

**Comment by alvestrand** [comment]
The spec already contains the verbiage "An RTCMediaStreamTrackStats object represents the stats about one attachment of a MediaStreamTrack to the PeerConnection object for which one calls getStats." - this was written prior to RTPSender/RTPReceiver, and can be said to pre-shadow the concepts.

The replaceTrack() operation does complicate the situation (as usual). If you do replaceTrack(), the change from track to sender will have an observable difference; the track-based stats would show the old track as "detached" and the new track as a new stats object, while a sender-based stats would change the trackIdentifier value and keep the counters increasing.

A proposal I'm sure nobody will really love: *add* the sender/receiver based stats (with the splits provided in #230), and consider moving the track stats to the "obsolete" sections once we have implementation experience saying that the new objects serve the purposes better.

This means that for a while, we'll report all numbers twice. Not sure if that's a Big Deal or not.

**Comment by jan-ivar** [comment]
This would be simpler, and remove the need for [detached](https://w3c.github.io/webrtc-stats/#dom-rtcmediastreamtrackstats-detached), and accumulation of stale stats objects.

This removes the risk of users accidentally relying on stale stats from (detached) objects, when they either forget to filter them out properly, or combine the data incorrectly with the still-live version.

Firefox does not implement track stats yet. We would like to move directly to sender/receiver stats.

**Comment by alvestrand** [comment]
Decision from TPAC: Add "sender" and "receiver" stats, keep "track" for the moment, raise a separate issue about removing "track".

**Comment by alvestrand** [comment]
Solved by #273
